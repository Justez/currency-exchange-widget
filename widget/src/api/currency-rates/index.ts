import axios from 'axios'

import { Currencies,CurrencyRateResponse, CurrencyRateError } from 'types';

const URL = 'https://openexchangerates.org/api/latest.json';

export const getCurrencyRates = (obj: Currencies) => axios.get(URL, {
    params: {
        app_id: '2b3e9170d55c46fead589f4fdeac1fb8',
        // base: obj.out.toUpperCase(),
    }
})
    .then((res: CurrencyRateResponse) => res.data)
    .catch((err: CurrencyRateError) => err);