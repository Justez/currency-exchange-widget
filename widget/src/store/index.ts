import { createStore, applyMiddleware } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';

import createRootReducer from './root-reducer';
import sagas from './root-saga';

const configureStore = () => {
  const sagaMiddleware = createSagaMiddleware();

  const middlewares = [
    sagaMiddleware,
  ];

  const instance = createStore(
    createRootReducer(),
    composeWithDevTools(applyMiddleware(...middlewares)),
  );

  sagaMiddleware.run(sagas);

  return instance;
};

export type State = ReturnType<typeof store.getState>;

const store = configureStore();

export default store;
