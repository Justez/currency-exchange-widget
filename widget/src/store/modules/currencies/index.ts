//@ts-ignore
import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencies';

export const actions = createActions(
  'SET_SELECTED_CURRENCY',
  'FLIP_SELECTED_CURRENCIES',
  { prefix: NAMESPACE },
);
