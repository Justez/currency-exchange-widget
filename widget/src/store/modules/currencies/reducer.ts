import { handleActions } from 'redux-actions';
import { assocPath } from 'ramda';

import { Currencies, CurrencyMap } from 'types';

import { actions, NAMESPACE } from '.';

export interface DefaultState {
  currencies: Currencies;
}

export const defaultState: DefaultState = {
  currencies: {
    in: CurrencyMap.euro,
    out: CurrencyMap.pound,
  },
};

type Payload = any;

export const initializedState = {};

const reducer = handleActions<DefaultState, Payload>({
  [actions.setSelectedCurrency.toString()]: (state, { payload }) =>
    assocPath([NAMESPACE, Object.keys(payload)[0]], Object.values(payload)[0], state)
},
  defaultState,
);

export default reducer;
