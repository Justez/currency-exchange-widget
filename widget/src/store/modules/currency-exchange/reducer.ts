import { handleActions } from 'redux-actions';
// import { assocPath } from 'ramda';

// import type { rates } from 'types/api';

// import { actions } from '.';

export interface DefaultState {
  // currencyRates: rates;
}

export const defaultState: DefaultState = {
  currencyRates: null,
};

type Payload = any;

export const initializedState = {};

const reducer = handleActions<DefaultState, Payload>({
  // [actions.setCurrencyRates.toString()]: (state, { payload }) =>
  //     assocPath(['currencyRates'], payload.payload, state),
},
  defaultState,
);

export default reducer;
