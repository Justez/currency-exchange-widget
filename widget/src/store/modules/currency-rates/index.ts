import { createActions } from 'redux-actions';

export const NAMESPACE = 'currencyRates';

export const actions = createActions(
  'FLIP_RATES',
  'GET_CURRENCY_RATES',
  'GET_CURRENCY_RATES_REQUEST',
  'GET_CURRENCY_RATES_SUCCESS',
  'GET_CURRENCY_RATES_FAILURE',
  'SET_CURRENCY_RATES',
  'RESET_RATES',
  { prefix: NAMESPACE },
);
