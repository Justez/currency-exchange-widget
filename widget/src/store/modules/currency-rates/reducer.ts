import { handleActions } from 'redux-actions';
import { assocPath } from 'ramda';

import { CurrencyRates } from 'types';

import { actions } from '.';

export type DefaultState = CurrencyRates;

export const defaultState: DefaultState = {
  rate: '',
  reverse: '',
};

type Payload = any;

export const initializedState = {};

const reducer = handleActions<DefaultState, Payload>({
  [actions.setCurrencyRates.toString()]: (state, { payload }) =>
    assocPath([], { ...state, ...payload }, state),
  [actions.flipRates.toString()]: ((state) =>
    assocPath([], { rate: state.reverse, reverse: state.rate }, state)),
  [actions.resetRates.toString()]: ((state) =>
    assocPath([], defaultState, state)),
},
  defaultState,
);

export default reducer;
