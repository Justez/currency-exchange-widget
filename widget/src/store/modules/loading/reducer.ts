import { assoc, dissoc } from 'ramda';
import { AnyAction } from 'redux';

export interface DefaultState {
  [key: string]: boolean,
}

export const defaultState: DefaultState = {};

const reducer = (state: DefaultState = defaultState, action: AnyAction): DefaultState => {
  const { type } = action;
  const matches = /(.*)_(REQUEST|SUCCESS|FAILURE)/.exec(type);

  if (!matches) {
    return state;
  }

  const [, requestName, requestState] = matches;
  
  const isRequestActive = requestState === 'REQUEST';

  if (isRequestActive) {
    return assoc(requestName, isRequestActive, state);
  } else {
    return dissoc(requestName, state);
  }
};

export default reducer;
