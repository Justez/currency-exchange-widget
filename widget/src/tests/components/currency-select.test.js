import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';

import store from 'store';
import { CurrencyExchangeTypes, CurrencyMap } from 'types';
import { actions } from 'store/modules/currency-rates';
import { defaultState } from 'store/modules/currencies/reducer';
import { actions as currencyActions } from 'store/modules/currencies';
import CurrencySelect from 'components/currency-select-field';

const component = (
    <Provider store={store}>
        <CurrencySelect pocketDirection={CurrencyExchangeTypes.out} />
        <CurrencySelect pocketDirection={CurrencyExchangeTypes.in} />
    </Provider>
);

let wrapper;
let firstInput, secondInput;

describe('Currency Select integration', () => {
    beforeAll(() => {
        store.dispatch(actions.setCurrencyRates({ rate: 1.22, reverse: 0.88 }));
        wrapper = mount(component);
        firstInput = wrapper.find('#choose-out-currency').first();
        secondInput = wrapper.find('#choose-in-currency').first();
    });

    it('both selects are rendered', () => {
        expect(firstInput.exists()).toBeTruthy();
        expect(secondInput.exists()).toBeTruthy();
    });

    it('currency-out select shows default currency', () => {
        expect(firstInput.props().value).toBe(defaultState[CurrencyExchangeTypes.out]);
        expect(secondInput.props().value).toBe(defaultState[CurrencyExchangeTypes.in]);
    });

    it('currency-out select change is visible after action', () => {
        store.dispatch(currencyActions.setSelectedCurrency({ [CurrencyExchangeTypes.out]: CurrencyMap.dollar }));
        wrapper = mount(component);
        firstInput = wrapper.find('#choose-out-currency').first();
        expect(firstInput.props().value).toBe(CurrencyMap.dollar);
    });

    it('currency-in select change is visible after action', () => {
        store.dispatch(currencyActions.setSelectedCurrency({ [CurrencyExchangeTypes.in]: CurrencyMap.euro }));
        wrapper = mount(component);
        secondInput = wrapper.find('#choose-in-currency').first();
        expect(secondInput.props().value).toBe(CurrencyMap.euro);
    });
});

describe('CurrencySelect snapshots', () => {
    it('renders selects', () => {
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders selects after store change', () => {
        act(() => {
            store.dispatch(currencyActions.setSelectedCurrency({ [CurrencyExchangeTypes.in]: CurrencyMap.pound }));
            store.dispatch(currencyActions.setSelectedCurrency({ [CurrencyExchangeTypes.out]: CurrencyMap.euro }));
        });

        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
    });
});


export default component;