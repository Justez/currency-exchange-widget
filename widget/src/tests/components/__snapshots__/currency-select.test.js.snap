// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencySelect snapshots renders selects 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled"
        onClick={[Function]}
      >
        <div
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-labelledby="choose-out-currency"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
          id="choose-out-currency"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={null}
          role="button"
          tabIndex={null}
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            USD
          </h5>
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="usd"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <p
        className="MuiTypography-root makeStyles-balance-1 MuiTypography-body2"
      >
        Balance
         
        7699.22
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "height": "16px",
              "marginBottom": "-2px",
              "marginLeft": "-4px",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M11.8 10.9c-2.27-.59-3-1.2-3-2.15 0-1.09 1.01-1.85 2.7-1.85 1.78 0 2.44.85 2.5 2.1h2.21c-.07-1.72-1.12-3.3-3.21-3.81V3h-3v2.16c-1.94.42-3.5 1.68-3.5 3.61 0 2.31 1.91 3.46 4.7 4.13 2.5.6 3 1.48 3 2.41 0 .69-.49 1.79-2.7 1.79-2.06 0-2.87-.92-2.98-2.1h-2.2c.12 2.19 1.76 3.42 3.68 3.83V21h3v-2.15c1.95-.37 3.5-1.5 3.5-3.55 0-2.84-2.43-3.81-4.7-4.4z"
          />
        </svg>
      </p>
    </div>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled"
        onClick={[Function]}
      >
        <div
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-labelledby="choose-in-currency"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
          id="choose-in-currency"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={null}
          role="button"
          tabIndex={null}
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            EUR
          </h5>
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="eur"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <p
        className="MuiTypography-root MuiTypography-body2"
      >
        Balance
         
        868.22
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "height": "15px",
              "marginBottom": "-2px",
              "width": "15px",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M15 18.5c-2.51 0-4.68-1.42-5.76-3.5h5.14c.38 0 .73-.21.89-.55.33-.66-.15-1.45-.89-1.45h-5.8c-.05-.33-.08-.66-.08-1s.03-.67.08-1h5.8c.38 0 .73-.21.89-.55.34-.67-.14-1.45-.89-1.45H9.24C10.32 6.92 12.5 5.5 15 5.5c1.25 0 2.42.36 3.42.97.5.31 1.15.26 1.57-.16.58-.58.45-1.53-.25-1.96C18.36 3.5 16.73 3 15 3c-3.92 0-7.24 2.51-8.48 6h-2.9c-.38 0-.73.21-.9.55-.33.67.15 1.45.9 1.45h2.44c-.04.33-.06.66-.06 1s.02.67.06 1H3.62c-.38 0-.73.21-.89.55-.34.67.14 1.45.89 1.45h2.9c1.24 3.49 4.56 6 8.48 6 1.74 0 3.36-.49 4.74-1.35.69-.43.82-1.39.24-1.97-.42-.42-1.07-.47-1.57-.15-.99.62-2.15.97-3.41.97z"
          />
        </svg>
      </p>
    </div>
  </div>,
]
`;

exports[`CurrencySelect snapshots renders selects after store change 1`] = `
Array [
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled"
        onClick={[Function]}
      >
        <div
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-labelledby="choose-out-currency"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
          id="choose-out-currency"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={null}
          role="button"
          tabIndex={null}
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            EUR
          </h5>
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="eur"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <p
        className="MuiTypography-root makeStyles-balance-1 MuiTypography-body2"
      >
        Balance
         
        868.22
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "height": "15px",
              "marginBottom": "-2px",
              "width": "15px",
            }
          }
          viewBox="0 0 24 24"
        >
          <path
            d="M15 18.5c-2.51 0-4.68-1.42-5.76-3.5h5.14c.38 0 .73-.21.89-.55.33-.66-.15-1.45-.89-1.45h-5.8c-.05-.33-.08-.66-.08-1s.03-.67.08-1h5.8c.38 0 .73-.21.89-.55.34-.67-.14-1.45-.89-1.45H9.24C10.32 6.92 12.5 5.5 15 5.5c1.25 0 2.42.36 3.42.97.5.31 1.15.26 1.57-.16.58-.58.45-1.53-.25-1.96C18.36 3.5 16.73 3 15 3c-3.92 0-7.24 2.51-8.48 6h-2.9c-.38 0-.73.21-.9.55-.33.67.15 1.45.9 1.45h2.44c-.04.33-.06.66-.06 1s.02.67.06 1H3.62c-.38 0-.73.21-.89.55-.34.67.14 1.45.89 1.45h2.9c1.24 3.49 4.56 6 8.48 6 1.74 0 3.36-.49 4.74-1.35.69-.43.82-1.39.24-1.97-.42-.42-1.07-.47-1.57-.15-.99.62-2.15.97-3.41.97z"
          />
        </svg>
      </p>
    </div>
  </div>,
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 MuiGrid-direction-xs-column"
  >
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled"
        onClick={[Function]}
      >
        <div
          aria-disabled="true"
          aria-haspopup="listbox"
          aria-labelledby="choose-in-currency"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled Mui-disabled"
          id="choose-in-currency"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={null}
          role="button"
          tabIndex={null}
        >
          <h5
            className="MuiTypography-root MuiTypography-h5"
          >
            GBP
          </h5>
        </div>
        <input
          onAnimationStart={[Function]}
          type="hidden"
          value="gbp"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSelect-icon Mui-disabled"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <p
        className="MuiTypography-root MuiTypography-body2"
      >
        Balance
         
        102.22
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root"
          focusable="false"
          style={
            Object {
              "height": "11",
              "marginLeft": "4px",
              "width": "11",
            }
          }
          viewBox="0 0 8 8"
        >
          <path
            d="M3 0c-.62 0-1.16.26-1.5.69-.34.43-.5.99-.5 1.56 0 .69.16 1.25.25 1.75h-1.25v1h1.22c-.11.45-.37.96-1.06 1.66l-.16.13v1.22h6v-1h-4.91c.64-.73.98-1.4 1.13-2h1.78v-1h-1.72c-.08-.68-.28-1.24-.28-1.75 0-.39.11-.73.28-.94.17-.21.37-.31.72-.31.39 0 .61.11.75.25s.25.36.25.75h1c0-.58-.17-1.1-.53-1.47-.37-.37-.89-.53-1.47-.53z"
            xmlns="http://www.w3.org/2000/svg"
          />
        </svg>
      </p>
    </div>
  </div>,
]
`;
